loyalty_criteria_multi_product/migrations/pre-migration.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2023 Tecnativa - Pilar Vargas
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openupgradelib import openupgrade
+
+
+def move_coupon_criteria_to_rule(env):
+    """Move everything related to coupon.criteria from the loyalty.program model to
+    loyalty.rule."""
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE loyalty_rule
+        ADD COLUMN IF NOT EXISTS coupon_criteria CHAR
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        INSERT INTO loyalty_rule (
+            program_id,
+            company_id,
+            create_uid,
+            write_uid,
+            product_domain,
+            reward_point_mode,
+            minimum_amount_tax_mode,
+            mode,
+            active,
+            reward_point_split,
+            create_date,
+            write_date,
+            reward_point_amount,
+            loyalty_criteria
+        )
+        SELECT
+            lp.id,
+            lp.company_id,
+            lp.create_uid,
+            lp.write_uid,
+            '{}',
+            'order',
+            'incl',
+            'auto',
+            lp.active,
+            false,
+            lp.create_date,
+            lp.write_date,
+            1,
+            'multi_product'
+        FROM
+            loyalty_program AS lp
+        WHERE
+            lp.rule_id IS NULL
+            AND lp.coupon_criteria = 'multi_product'
+        """,
+    )
+    # If there are rules defined for the program, the value of coupon_criteria shall be
+    # defined in those rules, in case there are no previous rules for that program, a
+    # record shall be created for a new related rule to define coupon_criteria.
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE loyalty_rule AS lr
+        SET loyalty_criteria = 'multi_product'
+        FROM loyalty_program AS lp
+        WHERE lr.program_id = lp.id
+        AND lp.coupon_criteria = 'multi_product'
+        """,
+    )
```

Comment:
How can this be if `loyalty_rule` is a new relation?

Reply:
The data corresponding to the rules are migrated to this table. 
https://github.com/OCA/OpenUpgrade/blob/5912a50fa93234fe37643c1091919f780d718dd1/openupgrade_scripts/scripts/loyalty/16.0.1.0/pre-migration.py#L39

Reply:
Then why is it necessary to create them in the previous query?

Reply:
I have checked my notes again. In v15 "coupon_criteria" was defined at program level (coupon.program), now it has to be moved to rule level (loyalty.rule).  
v15: 
https://github.com/OCA/sale-promotion/blob/15.0/coupon_criteria_multi_product/models/coupon_program.py#L9

v16: https://github.com/OCA/sale-promotion/blob/60b86377e3b6c4de62086a2f266c0769c8111bd7/loyalty_criteria_multi_product/models/loyalty_rule.py#L9

That program could have or not defined rules that were stored in the "coupon_rule" table. 
v15: https://github.com/odoo/odoo/blob/15.0/addons/coupon/models/coupon_rules.py#L12-L19

If a program already has defined a rule, then there is already a record for that program and you have to adapt the data of the rule to consider "loyalty_criteria" but in case that program did not have other rules defined then you have to insert a new record to define the rule. So when re-analysing, the order of the queries is not relevant.

However, I have corrected the nomenclature of the field in loyalty_rule. Also the field loyalty_program.coupon_criteria does not have to be named when a new field is created in loyalty_rule.loyalty_criteria.

Reply:
What I mean is that for every program there will be a rule as it's inherent to the way `coupon.rule` is binded to the `coupon.program` model. So when a program is created the rule is created and therefor it makes no sense to create a new one.

Reply:
Right, by default a record with the value "rule_min_quantity" is set even if there are no other rules set. I was thinking about the possibility of a program configured only with the multi-product criteria and that this would not affect the rules, but I have already verified that it does. Just update the record, no need to insert new records. Thank you.